### PR TITLE
docs: permissions & reference roles

### LINKED ISSUE #75: Easy way to list Azure and AWS needed permissions
Coming from [here](https://discord.com/channels/485586884165107732/728693131557732403/817377801232711691)

> what are the AWS permissions required by the terraform provider for cml?

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
Closes #75; may need extra eyes and opinions on documentation and some awkward `kubectl` scripting.

### Potential follow–ups
* [ ] Validate all this code as part of our CI/CD `smoke` test workflow[^1]
* [ ] Migrate Kubernetes YAML to HCL for consistency[^2]
* [ ] Provide tighter ABAC apart from basic RBAC permissions[^3]

> ### Exporting credentials (one–liner)
> ```console
> eval "$(terraform output --json | jq --raw-output 'to_entries[]|"export \(.key|ascii_upcase)=\(.value.value|@sh)"')"
> ```
> Note: added `eval` to support multi–line variables like `gcp` service account credentials.
> 
> _Originally posted by @0x2b3bfa0 in https://github.com/iterative/terraform-provider-iterative/issues/75#issuecomment-1057074357_

[^1]: Requires some internal infrastructure work to create the appropriate sandboxes.
[^2]: Doesn't seem possible without a https://github.com/hashicorp/terraform-provider-kubernetes/issues/1641
[^3]: https://twitter.com/michaelgat/status/1017166804139954176

### CODE REVIEW

**Comment by casperdcl** [comment]
I presume the `permissions` root folder is WIP? Would it be placed in `docs/`, `tests/`, `.github/workflows` or somewhere else?

**Comment by 0x2b3bfa0** [comment]
Either `tests/` or `examples/` could be a good shoehorn–in–hand fit. Using `docs/permissions/` might be also appropriate.

> I presume the `permissions` root folder is WIP?

If “WIP” stands for “**w**orth **i**m**p**roving” then yes. 😅

**Comment by casperdcl** [comment]
Yes in any case could you also add a page `docs/guides/permissions.md`?

**Comment by 0x2b3bfa0** [comment]
Then,
* Rename `permissions/README.md` to `docs/guides/permissions.md` 
* Rename `permissions` to `docs/permissions` ❓

**Review by casperdcl — CHANGES_REQUESTED** [review]
actually can we merge `permissions.md` with `authentication.md`? They shouldn't be separate pages IMO. The latter name is probably better to keep.

File: docs/guides/permissions.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+---
+page_title: Permissions
```
[the remark this answers is not in the document]
Comment (0x2b3bfa0) [reply docs/guides/permissions.md:2]: > _Development: https://dvc.org/doc/user-guide_

🔥 🐶

[the remark this answers is not in the document]
Comment (0x2b3bfa0) [reply docs/guides/permissions.md]: > _probably not for the main guide, right?_

Probably yes, if we merge it with `authentication.md` in the main menu.

File: docs/guides/authentication.md [hunk 2]
```diff
@@ -40,3 +40,42 @@ Either one of:
 
 - `KUBECONFIG` - Path to a [`kubeconfig` file](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig/#the-kubeconfig-environment-variable).
 - `KUBECONFIG_DATA` - Alternatively, the **contents** of a `kubeconfig` file.
+
+# Sample Permissions
+
+The [docs/permissions](https://github.com/iterative/terraform-provider-iterative/tree/master/docs/permissions) directory contains sample roles and permissions to use TPI in all the supported cloud providers.
```
Comment (casperdcl) [inline docs/guides/authentication.md:46]: Prob worth moving?

```suggestion
The [permissions](https://github.com/iterative/terraform-provider-iterative/tree/master/docs/guides/permissions) directory contains sample roles and permissions to use TPI in all the supported cloud providers.
```

File: docs/guides/authentication.md [hunk 3]
```diff
@@ -40,3 +40,42 @@ Either one of:
...
+- [`az` (Microsoft Azure)](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/azure_cli)
+- [`gcp` (Google Cloud)](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started)
+- [`k8s` (Kubernetes)](https://kubernetes.io/docs/concepts/configuration/organize-cluster-access-kubeconfig)
+
+## Setting up permissions, credentials and environment variables
+
+### `aws`
+
+- Run `terraform init` and `terraform apply` in the `aws` directory
+- Set the [`AWS_ACCESS_KEY_ID`](https://registry.terraform.io/providers/iterative/iterative/latest/docs#AWS_ACCESS_KEY_ID) environment variable to the value returned by `terraform output --raw aws_access_key_id`
```
Comment (casperdcl) [inline docs/guides/authentication.md:62]: update these URLs after #433

```suggestion
- Set the [`AWS_ACCESS_KEY_ID`](https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/authentication#AWS_ACCESS_KEY_ID) environment variable to the value returned by `terraform output --raw aws_access_key_id`
```

Or better?

```suggestion
- Set the [`AWS_ACCESS_KEY_ID`](#AWS_ACCESS_KEY_ID) environment variable to the value returned by `terraform output --raw aws_access_key_id`
```

File: docs/guides/authentication.md [hunk 4]
```diff
@@ -40,3 +40,42 @@ Either one of:
...
+- Set the [`AWS_SECRET_ACCESS_KEY`](https://registry.terraform.io/providers/iterative/iterative/latest/docs#AWS_SECRET_ACCESS_KEY) environment variable to the value returned by `terraform output --raw aws_secret_access_key`
+
+### `az`
+
+- Run `terraform init` and `terraform apply` in the `az` directory
+- Set the [`AZURE_TENANT_ID`](https://registry.terraform.io/providers/iterative/iterative/latest/docs#AZURE_TENANT_ID) environment variable to the value returned by `terraform output --raw azure_tenant_id`
+- Set the [`AZURE_SUBSCRIPTION_ID`](https://registry.terraform.io/providers/iterative/iterative/latest/docs#AZURE_SUBSCRIPTION_ID) environment variable to the value returned by `terraform output --raw azure_subscription_id`
+- Set the [`AZURE_CLIENT_ID`](https://registry.terraform.io/providers/iterative/iterative/latest/docs#AZURE_CLIENT_ID) environment variable to the value returned by `terraform output --raw azure_client_id`
+- Set the [`AZURE_CLIENT_SECRET`](https://registry.terraform.io/providers/iterative/iterative/latest/docs#AZURE_CLIENT_SECRET) environment variable to the value returned by `terraform output --raw azure_client_secret`
```
Comment (casperdcl) [inline docs/guides/authentication.md:71]: alternatively, codify the lot?

```suggestion
cd permissions/az
terraform init
terraform apply
export AZURE_TENANT_ID="$(terraform output --raw azure_tenant_id)"
export AZURE_SUBSCRIPTION_ID="$(terraform output --raw azure_subscription_id)"
export AZURE_CLIENT_ID="$(terraform output --raw azure_client_id)"
export AZURE_CLIENT_SECRET="$(terraform output --raw azure_client_secret)"
```

File: docs/permissions/aws/main.tf [hunk 5]
```diff
@@ -0,0 +1,87 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 4.3.0"
+    }
```
Comment (casperdcl) [inline docs/permissions/aws/main.tf:6]: afaik `terrafomr fmt` is fine with this compression, right?

```suggestion
    aws = { source  = "hashicorp/aws", version = "~> 4.3.0" }
```

**Review by casperdcl — CHANGES_REQUESTED** [review]
getting there!

**Comment by 0x2b3bfa0** [comment]
I took the liberty of merging both sections following a pretty free interpretation of your suggestions.

File: docs/permissions/aws/main.tf (hunk 5 above)
Comment (0x2b3bfa0) [reply docs/permissions/aws/main.tf]: 🤣

**Review by casperdcl — COMMENTED** [review]
minor

**Review by casperdcl — APPROVED** [review]
one slight concern I have is the syntax won't work on all shells (e.g. windows) but is probably understandable to the sort of users it targets.

File: docs/guides/permissions/aws/main.tf [hunk 6]
```diff
@@ -0,0 +1,87 @@
+terraform {
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 4.3.0"
+    }
```
Comment (casperdcl) [inline docs/guides/permissions/aws/main.tf:6]: what happened to the minification?

```suggestion
    aws = { source = "hashicorp/aws", version = "~> 4.3.0" }
```

Comment (0x2b3bfa0) [reply docs/guides/permissions/aws/main.tf]: 🤦🏼‍♂️ I never applied it. It was unindented, I swear!

**Review by casperdcl — APPROVED** [review]
lgtm.

(minor) could we remove as many of the empty newlines in the `*.tf`/`*.sh` files as possible?

**Comment by 0x2b3bfa0** [comment]
> * _When both arguments and blocks appear together inside a block body, place all of the arguments together at the top and then place nested blocks below them. Use one blank line to separate the arguments from the blocks [...]_
> * _Use empty lines to separate logical groups of arguments within a block_ [...]
> * _Top-level blocks should always be separated from one another by one blank line. Nested blocks should also be separated by blank lines_ [...]

(Infimum) Not at the cost of breaking the [Style Conventions](https://www.terraform.io/language/syntax/style)?

File: docs/guides/permissions/k8s/kubeconfig.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  export KUBECONFIG="$(mktemp)"
+  {
+    kubectl config set-cluster cluster --server="https://$SERVER"
+    kubectl config set clusters.cluster.certificate-authority-data "$AUTHORITY"
+    kubectl config set-credentials task --token="$TOKEN"
+    kubectl config set-context cluster --cluster=cluster --user=task
+    kubectl config use-context cluster
+  } >/dev/null
+  cat "$KUBECONFIG" && rm "$KUBECONFIG"
+)
```
Comment (0x2b3bfa0) [inline docs/guides/permissions/k8s/kubeconfig.sh:33]:
```suggestion
#!/bin/bash

SERVER="$(
  kubectl config view --raw --flatten --output \
    jsonpath='{.clusters[0].cluster.server}'
)"
AUTHORITY="$(
  kubectl config view --raw --flatten --output \
    jsonpath='{.clusters[0].cluster.certificate-authority-data}'
)"
SECRET="$(
  kubectl get serviceaccount task --output \
    jsonpath="{.secrets[0].name}"
)"
TOKEN="$(
  kubectl get secret "$SECRET" --output \
    jsonpath="{.data.token}" | base64 --decode
)"

(
  export KUBECONFIG="$(mktemp)"
  {
    kubectl config set-cluster cluster --server="https://$SERVER"
    kubectl config set clusters.cluster.certificate-authority-data "$AUTHORITY"
    kubectl config set-credentials task --token="$TOKEN"
    kubectl config set-context cluster --cluster=cluster --user=task
    kubectl config use-context cluster
  } >/dev/null
  cat "$KUBECONFIG" && rm "$KUBECONFIG"
)
```

May I leave the newline behind the shebang and the line before the last block?

Comment (casperdcl) [reply docs/guides/permissions/k8s/kubeconfig.sh:33]: if you insist :D

Comment (0x2b3bfa0) [reply docs/guides/permissions/k8s/kubeconfig.sh:33]: No opinions on this one. If you feel like they should be removed, let's remove it.

**Comment by casperdcl** [comment]
> Not at the cost of breaking the [Style Conventions](https://www.terraform.io/language/syntax/style)

I think those conventions were written without deep thought.

If `terraform fmt` is happy we should be too, no?

File: docs/guides/permissions/k8s/kubeconfig.sh (hunk 7 above)
Comment (0x2b3bfa0) [reply docs/guides/permissions/k8s/kubeconfig.sh:33]: As well, if you feel the need of compacting some blocks, it's fine.

**Comment by 0x2b3bfa0** [comment]
> _I think those conventions were written without deep thought._

[Deep Thought](https://en.wikipedia.org/wiki/List_of_The_Hitchhiker%27s_Guide_to_the_Galaxy_characters#Deep_Thought), I presume?

> _If `terraform fmt` is happy we should be too, no?_

Inversely, I believe that `terraform fmt` was _written without deep thought™_

**Comment by 0x2b3bfa0** [comment]
> _adding and removing newlines is relatively easier to do [compared to horizontal whitespace] and often done for subjective reasons that an automatic tool cannot reason about_
* https://github.com/hashicorp/terraform/issues/23223#issuecomment-547519852

File: docs/guides/permissions/k8s/kubeconfig.sh (hunk 7 above)
Comment (0x2b3bfa0) [reply docs/guides/permissions/k8s/kubeconfig.sh:33]:
```suggestion
#!/bin/bash
SERVER="$(
  kubectl config view --raw --flatten --output \
    jsonpath='{.clusters[0].cluster.server}'
)"
AUTHORITY="$(
  kubectl config view --raw --flatten --output \
    jsonpath='{.clusters[0].cluster.certificate-authority-data}'
)"
SECRET="$(
  kubectl get serviceaccount task --output \
    jsonpath="{.secrets[0].name}"
)"
TOKEN="$(
  kubectl get secret "$SECRET" --output \
    jsonpath="{.data.token}" | base64 --decode
)"
export KUBECONFIG="$(mktemp)"
{
  kubectl config set-cluster cluster --server="https://$SERVER"
  kubectl config set clusters.cluster.certificate-authority-data "$AUTHORITY"
  kubectl config set-credentials task --token="$TOKEN"
  kubectl config set-context cluster --cluster=cluster --user=task
  kubectl config use-context cluster
} > /dev/null
cat "$KUBECONFIG" && rm "$KUBECONFIG"
```

**Comment by 0x2b3bfa0** [comment]
## [XKCD 859](https://xkcd.com/859)

**Comment by 0x2b3bfa0** [comment]
I'd prefer not to remove these newlines myself, on principle. Still, a commit with that change would be welcome.

**Comment by casperdcl** [comment]
Hope b74a784 is fine.

About style, there are two options: 1) aim for comprehension-over-rules, or 2) use a comprehensive linter[^1]. Since as you say `terraform fmt` isn't quite (2) then I'd revert to (1) i.e. aid first-time readers 802da6a.

[^1]: e.g. I detest [black](https://github.com/psf/black) & [yapf](https://github.com/google/yapf) but tolerate them since at least they're comprehensive

**Comment by 0x2b3bfa0** [comment]
Yes, there is no canonical representation for HCL apart from @casperdcl–taste-format and @0x2b3bfa0-style-guide-format. In absence of machines there is no clear way of make a neutral choice and differences are minimal, so my pragmatic vote is for the former. 🏳️ 

`set -euxo pipefail` in b74a784d3d84a7ec0c94a1db9ef5f868de23e9be is great to have, although I don't know if users will be misled somehow by `set -x` to `stderr` 🤔 

---

_Post scriptum: I adore [black](https://github.com/psf/black)... --line-length=79._

**Comment by 0x2b3bfa0** [comment]
:shipit: Ready to merge unless you want to `sed '/^$/d'` any other file. 😄

**Pull request merged**